ghcide/src/Development/IDE/Plugin/ChangeTypeAction.hs

```diff
@@ -0,0 +1,130 @@
+                                                 type (|?) (InR))
+import           Text.Regex.TDFA                ((=~))
+
+descriptor :: PluginId -> PluginDescriptor IdeState
+descriptor plId = (defaultPluginDescriptor plId) { pluginHandlers = mkPluginHandler STextDocumentCodeAction codeActionHandler }
+
+codeActionHandler :: PluginMethodHandler IdeState 'TextDocumentCodeAction
+codeActionHandler ideState plId CodeActionParams {_textDocument = TextDocumentIdentifier uri, _context = CodeActionContext (List diags) _} = response $ do
+      nfp <- getNormalizedFilePath plId (TextDocumentIdentifier uri)
+      decls <- getDecls ideState nfp
+      let actions = generateActions uri decls diags
+      pure $ List actions
+
+getDecls :: MonadIO m => IdeState -> NormalizedFilePath -> ExceptT String m [LHsDecl GhcPs]
+getDecls state = handleMaybeM "Error: Could not get Parsed Module"
+    . liftIO
+    . fmap (fmap (hsmodDecls . unLoc . pm_parsed_source))
+    . runAction "changeSignature.GetParsedModule" state
+    . use GetParsedModule
+
+data ChangeSignature = ChangeSignature { expectedType :: Text
+                                       , actualType   :: Text
+                                       , declName     :: Text
+                                       , declSrcSpan  :: Maybe SrcSpan
+                                       , diagnostic   :: Diagnostic
+                                       , uri          :: Uri
+                                       }
+
+-- Needed to trackdown OccNames in signatures
+type SigName p = (HasOccName (IdP (GhcPass p)))
+
+-- | Generate CodeActions from a list of Diagnostics
+generateActions :: SigName p => Uri -> [LHsDecl (GhcPass p)] -> [Diagnostic] -> [Command |? CodeAction]
+generateActions uri = mapMaybe . generateAction uri
+
+-- | Create a CodeAction from a Diagnostic
+generateAction :: SigName p => Uri -> [LHsDecl (GhcPass p)] -> Diagnostic -> Maybe (Command |? CodeAction)
+generateAction uri decls diag = diagnosticToChangeSig uri decls diag >>= changeSigToCodeAction
+
+-- | Convert a diagnostic into a ChangeSignature and add the proper SrcSpan
+diagnosticToChangeSig :: SigName p => Uri -> [LHsDecl (GhcPass p)] -> Diagnostic -> Maybe ChangeSignature
+diagnosticToChangeSig uri decls diag = addSrcSpan decls <$> matchingDiagnostic uri diag
+
+-- | If a diagnostic has the proper message create a ChangeSignature from it
+matchingDiagnostic :: Uri -> Diagnostic -> Maybe ChangeSignature
+matchingDiagnostic uri diag@Diagnostic{_message} = unwrapMatch $ _message =~ expectedMessage
+    where
+        expectedMessage = "Expected type: (.+)\n +Actual type: (.+)\n.*\n +In an equation for ‘(.+)’" :: Text
+        unwrapMatch :: (Text, Text, Text, [Text]) -> Maybe ChangeSignature
+        unwrapMatch (_, _, _, [exp, act, name]) = Just $ ChangeSignature exp act name Nothing diag uri
+        unwrapMatch _                           = Nothing
```

Comment:
This looks fine, but GHC 9.4 offers a typed api for diagnostics which hopefully removes the need to parse error messages. See #2014

Reply:
Do you think it would be better to hold off until the new Diagnostic changes are integrated? Reading through the issue it seems like a big effort so maybe it would be better just to start now and migrate to the new format down the road?

Reply:
No need to wait, we still want present GHCs to be able to use this plugin